timemachine/cpp/src/kernels/k_nonbonded_pairs.cuh

```diff
@@ -0,0 +1,208 @@
+
+    RealType real_cutoff = static_cast<RealType>(cutoff);
+    RealType cutoff_squared = real_cutoff * real_cutoff;
+
+    RealType charge_scale = scales[pair_idx * 2 + 0];
+    RealType lj_scale = scales[pair_idx * 2 + 1];
+
+    RealType box_x = box[0 * 3 + 0];
+    RealType box_y = box[1 * 3 + 1];
+    RealType box_z = box[2 * 3 + 2];
+
+    RealType inv_box_x = 1 / box_x;
+    RealType inv_box_y = 1 / box_y;
+    RealType inv_box_z = 1 / box_z;
+
+    RealType delta_x = ci_x - cj_x;
+    RealType delta_y = ci_y - cj_y;
+    RealType delta_z = ci_z - cj_z;
+
+    delta_x -= box_x * nearbyint(delta_x * inv_box_x);
+    delta_y -= box_y * nearbyint(delta_y * inv_box_y);
+    delta_z -= box_z * nearbyint(delta_z * inv_box_z);
+
+    RealType delta_w = ci_w - cj_w;
+    RealType d2ij = delta_x * delta_x + delta_y * delta_y + delta_z * delta_z + delta_w * delta_w;
+
+    unsigned long long energy = 0;
+
+    int is_vanilla =
+        (ci_w == 0 && dq_dl_i == 0 && dsig_dl_i == 0 && deps_dl_i == 0 && cj_w == 0 && dq_dl_j == 0 && dsig_dl_j == 0 &&
+         deps_dl_j == 0);
+
+    // see note: this must be strictly less than
+    if (d2ij < cutoff_squared) {
+
+        RealType u;
+        RealType ebd;
+        RealType es_prefactor;
+        RealType dij;
+        RealType inv_dij;
+        RealType inv_d2ij;
+        compute_electrostatics<RealType, true>(
+            charge_scale, qi, qj, d2ij, beta, dij, inv_dij, inv_d2ij, ebd, es_prefactor, u);
+
+        RealType delta_prefactor = es_prefactor;
+        RealType real_du_dl = 0;
+        // lennard jones force
+        if (eps_i != 0 && eps_j != 0) {
+            RealType sig_grad;
+            RealType eps_grad;
+            compute_lj<RealType, true>(
+                lj_scale, eps_i, eps_j, sig_i, sig_j, inv_dij, inv_d2ij, u, delta_prefactor, sig_grad, eps_grad);
+
+            g_sigi += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DSIG>(-sig_grad);
+            g_sigj += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DSIG>(-sig_grad);
+            g_epsi += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DEPS>(-eps_grad * eps_j);
+            g_epsj += FLOAT_TO_FIXED_DU_DP<RealType, FIXED_EXPONENT_DU_DEPS>(-eps_grad * eps_i);
+
+            real_du_dl -= sig_grad * (dsig_dl_i + dsig_dl_j);
+            RealType term = eps_grad * fix_nvidia_fmad(eps_j, deps_dl_i, eps_i, deps_dl_j);
+            real_du_dl -= term;
+        }
+
+        gi_x -= FLOAT_TO_FIXED_NONBONDED(delta_prefactor * delta_x);
+        gi_y -= FLOAT_TO_FIXED_NONBONDED(delta_prefactor * delta_y);
+        gi_z -= FLOAT_TO_FIXED_NONBONDED(delta_prefactor * delta_z);
+
+        gj_x -= FLOAT_TO_FIXED_NONBONDED(-delta_prefactor * delta_x);
+        gj_y -= FLOAT_TO_FIXED_NONBONDED(-delta_prefactor * delta_y);
+        gj_z -= FLOAT_TO_FIXED_NONBONDED(-delta_prefactor * delta_z);
+
+        // energy is size extensive so this may not be a good idea
+        energy -= FLOAT_TO_FIXED_NONBONDED(u);
```

Comment:
TODO: the implementation here is still specialized to exclusions (returns the _negative_ sum of pair potentials). We'll probably want flip the convention and return the positive potential since we intend to use this as a generic nonbonded pairs potential.